src/compiler/transformers/declarations/emitBinder.ts

```diff
@@ -324,21 +342,25 @@ export function bindSourceFileForDeclarationEmit(file: SourceFile) {
                 }
             }
             symbol.declarations.push(node);
-            symbol.flags |= includes;
             setValueDeclaration(symbol as Symbol, node);
+            symbol.flags |= includes;
             getNodeLinks(node).symbol = symbol;
             // Some parts of declarations.ts use the symbol directly. We provide enough of it for everything to work.
             node.symbol = symbol as Symbol;
             return symbol;
         }
         function withScope(scope: Node, exports: EmitDeclarationSymbolTable | undefined, fn: () => void) {
-            const old = [currentScope, currentLocalSymbolTable, currentExportsSymbolTable] as const;
+            const old = [currentScope, currentSymbol, currentFunctionLocalSymbolTable, currentLocalSymbolTable, currentExportsSymbolTable] as const;
             currentScope = scope;
             const links = getNodeLinks(scope);
             currentLocalSymbolTable = links.locals ??= new Map();
+            currentSymbol = links.symbol ?? currentSymbol;
             currentExportsSymbolTable = exports;
+            if (isBlockScopedContainerTopLevel(scope)) {

```

Comment:
Can you explain a bit what the effect of this and related code separating a function local table? I don't think I fully understand the flow of what changed here.

Reply:
Initially we didn't go into blocks however I recently found that TS will add expando members from nested blocks. Ex:

```ts
function Foo() {}

if(Math.random()) { Foo.test = 1; }
{ Foo. bar =2; }
```

This means that we now do care about variables defined in blocks and the fact that they are only scoped to that block since that means any member accesses in that block will not add expando members to any variable that is shadowed in that scope:


```ts
function Foo() {}

if(Math.random()) { Foo.test = 1; }
{ let Foo: any = {}; Foo. bar =2; } // bar not on Foo

```

[Playground Link](https://www.typescriptlang.org/play?ts=5.4.0-dev.20231205#code/GYVwdgxgLglg9mABAMTnAFASkQbwL4CwAUMTMOgLICGUAFgHQBOVYAJnALZbY4pr1QApgGcoiALyIAjAG5EhIrwA2gsajgAuRCwCeE3Hjnr6iAEZVGEgExy8QA)

(TS also handled this incorrectly until recently: https://github.com/microsoft/TypeScript/issues/56538)